zcash_test_vectors/zip_0244_zsa.py

```diff
@@ -0,0 +1,272 @@
+            if len(t_inputs) > 0:
+                txin = rand.a(t_inputs)
+            else:
+                txin = None
+
+            sighash_shielded = signature_digest(tx, t_inputs, SIGHASH_ALL, None)
+            other_sighashes = {
+                nHashType: None if txin is None else signature_digest(tx, t_inputs, nHashType, txin)
+                for nHashType in ([
+                                      SIGHASH_ALL,
+                                      SIGHASH_NONE,
+                                      SIGHASH_SINGLE,
+                                      SIGHASH_ALL | SIGHASH_ANYONECANPAY,
+                                      SIGHASH_NONE | SIGHASH_ANYONECANPAY,
+                                      SIGHASH_SINGLE | SIGHASH_ANYONECANPAY,
+                                      ] if txin is None or txin.nIn < len(tx.vout) else [
+                    SIGHASH_ALL,
+                    SIGHASH_NONE,
+                    SIGHASH_ALL | SIGHASH_ANYONECANPAY,
+                    SIGHASH_NONE | SIGHASH_ANYONECANPAY,
+                    ])
+            }
+
+            test_vectors.append({
+                'tx': bytes(tx),
+                'txid': txid,
+                'auth_digest': auth,
+                'amounts': [x.amount for x in t_inputs],
+                'script_pubkeys': [x.scriptPubKey.raw() for x in t_inputs],
+                'transparent_input': None if txin is None else txin.nIn,
+                'sighash_shielded': sighash_shielded,
+                'sighash_all': other_sighashes.get(SIGHASH_ALL),
+                'sighash_none': other_sighashes.get(SIGHASH_NONE),
+                'sighash_single': other_sighashes.get(SIGHASH_SINGLE),
+                'sighash_all_anyone': other_sighashes.get(SIGHASH_ALL | SIGHASH_ANYONECANPAY),
+                'sighash_none_anyone': other_sighashes.get(SIGHASH_NONE | SIGHASH_ANYONECANPAY),
+                'sighash_single_anyone': other_sighashes.get(SIGHASH_SINGLE | SIGHASH_ANYONECANPAY),
+            })
+
+    render_tv(
+        args,
+        'zip_0244',
+        (
+            ('tx',                    {'rust_type': 'Vec<u8>', 'bitcoin_flavoured': False}),
+            ('txid',                  '[u8; 32]'),
+            ('auth_digest',           '[u8; 32]'),
+            ('amounts',               'Vec<i64>'),
+            ('script_pubkeys',        {'rust_type': 'Vec<Vec<u8>>', 'bitcoin_flavoured': False}),
+            ('transparent_input',     'Option<u32>'),
+            ('sighash_shielded',      '[u8; 32]'),
+            ('sighash_all',           'Option<[u8; 32]>'),
+            ('sighash_none',          'Option<[u8; 32]>'),
+            ('sighash_single',        'Option<[u8; 32]>'),
+            ('sighash_all_anyone',    'Option<[u8; 32]>'),
+            ('sighash_none_anyone',   'Option<[u8; 32]>'),
+            ('sighash_single_anyone', 'Option<[u8; 32]>'),
+        ),
+        test_vectors,
+    )
+
+
+if __name__ == '__main__':
```

Comment:
generally speaking, total duplication does not work. Why not ,for example, reuse `def orchard_auth_digest(tx):` instead of def `orchard_zsa_auth_digest(tx: TransactionZSA):`. if needed you can introduce more parameters to the existing functions. But for some cases it is not required since (after var rename) both tx classes have similar internal structure.

Reply:
Introducing parameters worked in a lot of places, thanks.